File: app/src/main/java/com/revosleap/samplemusicplayer/utils/RecyclerAdapter.java
package com.revosleap.samplemusicplayer.utils;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.revosleap.samplemusicplayer.R;
import com.revosleap.samplemusicplayer.models.Song;

import java.util.ArrayList;

public class RecyclerAdapter extends RecyclerView.Adapter<RecyclerAdapter.ViewHolder> {
    public RecyclerAdapter(SongClicked clicked) {
        songClicked = clicked;
    }

    private ArrayList songsList = new ArrayList<Song>();
    private SongClicked songClicked;

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        return new ViewHolder(LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.track_item, viewGroup, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder viewHolder, int i) {
        final Song song= (Song) songsList.get(i);
        viewHolder.bind(song);
        viewHolder.itemView.setOnClickListener(
                new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        songClicked.onSongClicked(song);
                    }
                }
        );
    }

    @Override
    public int getItemCount() {
        return songsList.size();
    }

    public void addSongs(ArrayList songs) {
        songsList.clear();
        songsList = songs;
        notifyDataSetChanged();
    }

    public interface SongClicked {
        void onSongClicked(Song song);
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        private TextView title,artist;
        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            title = itemView.findViewById(R.id.textViewSongTitle);
            artist = itemView.findViewById(R.id.textViewArtistName);

        }
        void bind(Song song){
            title.setText(song.title);
            artist.setText(song.artistName);
        }
    }
}

File: app/src/main/java/com/revosleap/samplemusicplayer/utils/RecyclerAdapter.kt
package com.revosleap.samplemusicplayer.utils

import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.revosleap.samplemusicplayer.R
import com.revosleap.samplemusicplayer.models.Song

class RecyclerAdapter :
        androidx.recyclerview.widget.RecyclerView.Adapter<RecyclerAdapter.ViewHolder>() {
    private var onLongClick: OnLongClick? = null
    private var onSongClicked: SongClicked? = null
    private var songsSelected: SongsSelected? = null
    private var songsList = mutableListOf<Song>()
    private var selectedSongs = mutableListOf<Song>()
    private var selectionModeActive = false

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): ViewHolder {
        return ViewHolder(LayoutInflater.from(viewGroup.context).inflate(R.layout.track_item, viewGroup, false))
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        val song = songsList[position]
        viewHolder.bind(song, position)
        viewHolder.mainItem.isSelected = selectedSongs.contains(song)
        viewHolder.mainItem.setOnLongClickListener {
            onLongClick?.onSongLongClicked(position)
            if (!selectionModeActive) {
                selectionModeActive = true
            }
            false
        }
        viewHolder.mainItem.setOnClickListener {
            if (!selectionModeActive) {
                onSongClicked?.onSongClicked(song)
            } else {
                if (selectedSongs.contains(song)) {
                    selectedSongs.remove(song)
                    songsSelected?.onSelectSongs(getSelectedSongs())

                } else {
                    selectedSongs.add(song)

                    songsSelected?.onSelectSongs(getSelectedSongs())

                }
                notifyItemChanged(position)
            }
        }
    }

    override fun getItemCount(): Int {
        return songsList.size
    }

    fun addSongs(songs: MutableList<Song>) {
        songsList.clear()
        songsList.addAll(songs)
        notifyDataSetChanged()
    }

    fun removeSelection() {
        selectionModeActive = false
        selectedSongs.clear()
        notifyDataSetChanged()
    }

    fun getSelectedSongs(): MutableList<Song> {
        return selectedSongs
    }

    fun updateRemoved(song: Song){
        songsList.remove(song)
        notifyDataSetChanged()
    }

    fun setOnSongClicked(songClick: SongClicked) {
        this.onSongClicked = songClick
    }

    fun setOnLongClick(longClick: OnLongClick) {
        this.onLongClick = longClick
    }

    fun setSongsSelected(selection: SongsSelected) {
        this.songsSelected = selection
    }

    interface SongsSelected {
        fun onSelectSongs(selectedSongs: MutableList<Song>)
    }

    interface SongClicked {
        fun onSongClicked(song: Song)
    }

    interface OnLongClick {
        fun onSongLongClicked(position: Int)
    }

    inner class ViewHolder(itemView: View) : androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView) {
        private val title: TextView = itemView.findViewById(R.id.textViewSongTitle)
        private val artist: TextView = itemView.findViewById(R.id.textViewArtistName)
        val mainItem: ConstraintLayout = itemView.findViewById(R.id.mainConstraint)
        private var position: Int? = null

        fun bind(song: Song, pos: Int) {
            title.text = song.title
            artist.text = song.artistName
            position = pos

        }


    }
}
